relay-dynamic-config/src/metrics.rs

```diff
@@ -170,13 +171,113 @@ impl TransactionMetricsConfig {
+        Self { global, project }
+    }
+
+    /// Returns an iterator of metric specs.
+    pub fn metrics(&self) -> impl Iterator<Item = &MetricSpec> {
+        let project = self.project.metrics.iter();
+        let enabled_global = self
+            .enabled_templates()
+            .flat_map(|template| template.metrics.iter());
+
+        project.chain(enabled_global)
+    }
+
+    /// Returns an iterator of tag mappings.
+    pub fn tags(&self) -> impl Iterator<Item = &TagMapping> {
+        let project = self.project.tags.iter();
+        let enabled_global = self
+            .enabled_templates()
+            .flat_map(|template| template.tags.iter());
+
+        project.chain(enabled_global)
+    }
+
+    fn enabled_templates(&self) -> impl Iterator<Item = &MetricExtractionTemplate> {
+        self.global.templates.iter().filter_map(|(key, template)| {
+            let is_enabled_by_override =
+                self.project.global_templates.get(key).map(|c| c.is_enabled);
+            let is_enabled = is_enabled_by_override.unwrap_or(template.is_enabled);
+
+            is_enabled.then_some(template)
+        })
+    }
+}
+
+impl<'a> From<&'a MetricExtractionConfig> for CombinedMetricExtractionConfig<'a> {
+    fn from(value: &'a MetricExtractionConfig) -> Self {
+        Self::new(MetricExtractionTemplates::EMPTY, value)
+    }
+}
+
+/// Global templates for metric extraction.
+///
+/// Templates can be enabled or disabled by project configs.
+#[derive(Clone, Default, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct MetricExtractionTemplates {
+    /// Mapping from template name to metrics specs & tags.
+    #[serde(skip_serializing_if = "BTreeMap::is_empty")]
+    pub templates: BTreeMap<String, MetricExtractionTemplate>,
+}
+
+impl MetricExtractionTemplates {
+    const EMPTY: &'static Self = &Self {
+        templates: BTreeMap::new(),
+    };
+
+    /// Returns `true` if the continaed templates are empty.
+    pub fn is_empty(&self) -> bool {
+        self.templates.is_empty()
+    }
+}
+
+/// Set of metrics & tags that can be enabled or disabled as a group.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct MetricExtractionTemplate {
+    /// Whether the set is enabled by default.
+    ///
+    /// Project configs can overwrite this flag to opt-in or out of a set.
+    pub is_enabled: bool,
+
+    /// A list of metric specifications to extract.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub metrics: Vec<MetricSpec>,
+
+    /// A list of tags to add to previously extracted metrics.
+    ///
+    /// These tags add further tags to a range of metrics. If some metrics already have a matching
+    /// tag extracted, the existing tag is left unchanged.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub tags: Vec<TagMapping>,
```

Comment:
`MetricSpec` already has a `tags` property, can we reuse that one?

Reply:
I guess we could write custom (de)serializers and save some memory, but probably not worth the memory benefit.